src/main/java/org/truffleruby/language/objects/MetaClassNode.java

```diff
@@ -42,7 +42,11 @@ public final RubyClass executeCached(Object value) {
     public abstract RubyClass execute(Node node, Object value);
 
     @Specialization(
-            guards = { "isSingleContext()", "object == cachedObject", "metaClass.isSingleton" },
+            guards = {
+                    "isSingleContext()",
+                    "object == cachedObject",
+                    "metaClass.isSingleton",
+                    "metaClass == object.getMetaClass()" },
```

Comment:
This adds a runtime check and makes this specialization not really worth it anymore. Instead, we could guard on the cached object.getMetaClass() being not <= IO, or more simply `!(cachedObject instanceof RubyIO)`, that would then only be tested on the slow path and not the fast path.

+ change the vm_set_class primitive to only accepts RubyIO objects

Reply:
Done.